include/envoy/stats/stats.h

```diff
@@ -20,39 +21,90 @@ class Instance;
 namespace Stats {
 
 /**
+ * General representation of a tag.
+ */
+struct Tag {
+  std::string name_;
+  std::string value_;
+};
+
+class TagExtractor {
+public:
+  virtual ~TagExtractor() {}
+
+  /**
+   * Identifier for this tag.
+   */
+  virtual std::string name() const PURE;
+
+  /**
+   * Updates the tag extracted name and the set of tags by extracting the tag represented by this
+   * TagExtractor. If the tag is not represented in the current tag_extracted_name, nothing will be
+   * modified.
+   * @param[in,out] tag_extracted_name name from which the tag will be removed if found to be
```

Comment:
I'd probably find it cleaner to make the argument immutable and just generate a new tag extracted name. I don't think this is on any hot path so no performance issues from an additional allocation.